File: backend/history-service/middleware/authMiddleware.js
const firebaseAdmin = require("firebase-admin");

// Initialize Firebase Admin if not already initialized.
// In case other files import authMiddleware before Admin is set up, guard it.
if (!firebaseAdmin.apps.length) {
  // We assume that `GOOGLE_APPLICATION_CREDENTIALS` was set in .env
  const serviceAccount = require("../firebase-service-account.json");
  firebaseAdmin.initializeApp({
    credential: firebaseAdmin.credential.cert(serviceAccount),
  });
}

async function verifyToken(req, res, next) {
  const authHeader = req.headers.authorization || "";
  // Header format: "Bearer <token>"
  const match = authHeader.match(/^Bearer (.+)$/);

  if (!match) {
    res.status(401).json({ error: "Missing or malformed Authorization header" });
    return;
  }

  const idToken = match[1];

  try {
    const decodedToken = await firebaseAdmin.auth().verifyIdToken(idToken);
    // decodedToken has { uid, email, name, ... }
    req.user = decodedToken;
    return next();
  } catch (err) {
    console.error("Error verifying ID token:", err);
    res.status(401).json({ error: "Invalid or expired ID token" });
    return;
  }
}

module.exports = verifyToken;

File: backend/message-analyser-service/middleware/authMiddleware.py
import os
import firebase_admin
from firebase_admin import credentials, auth
from fastapi import Request, HTTPException, status
from dotenv import load_dotenv

load_dotenv()

# Initialize Firebase Admin (if not already initialized)
if not firebase_admin._apps:
    cred_path = os.getenv("GOOGLE_APPLICATION_CREDENTIALS", "firebase-service-account.json")
    cred = credentials.Certificate(cred_path)
    firebase_admin.initialize_app(cred)

async def verify_token(request: Request):
    """
    Dependency to verify Firebase ID Token passed in 'Authorization: Bearer <token>'.
    On success, returns the decoded token (uid, email, etc.).
    On failure, raises HTTP 401.
    """
    auth_header = request.headers.get("Authorization")
    if not auth_header:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Missing Authorization header")

    scheme, _, token = auth_header.partition(" ")
    if scheme.lower() != "bearer" or not token:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid Authorization header format")

    try:
        decoded_token = auth.verify_id_token(token)
        return decoded_token  # contains 'uid', 'email', etc.
    except Exception as e:
        # Token is invalid or expired
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid or expired ID token")
